README.rst
===========================================
async_tasks - Lightweight async build tool
===========================================
microbuild fork

Added
========

* Async executing on worker
* Time of executing of each task
* Better logging


Rewrited
========
* Logging with case thread-safe
* Task executing


Removed
========
* Arg parsing and cl interface (will add soon)
* Ignore task


Example
========

::

    import os
    import shutil
    import zipfile

    import requests

    from async_tasks.async_tasks import task, run

    dir = './temp/'


    def load_file(url, name):
        response = requests.get(url, stream=True)
        with open(dir + name, "wb") as handle:
            handle.write(response.content)
        return dir + name


    @task()
    def create_temp_dir(logger):
        if not os.path.exists(dir):
            logger("create temp dir")
            os.makedirs(dir)
        else:
            logger("clear temp dir")
            shutil.rmtree(dir)
            os.makedirs(dir)


    @task(create_temp_dir)
    def load_logo(logger):
        file_address = load_file("https://avatars1.githubusercontent.com/u/13404754?v=3&s=460", "punksta_logo.png")
        return file_address


    @task(create_temp_dir)
    def load_content(logger):
        file_address = load_file("https://raw.githubusercontent.com/punksta/async_tasks/master/example.py", "example.py")
        return file_address


    @task(load_content, load_logo)
    def log_result(logger):
        for name in [load_content.result, load_logo.result]:
            logger(name + " loaded")


    @task(load_content, load_logo)
    def zip_result(logger):
        zf = zipfile.ZipFile(dir + 'result.zip', mode='w')

        logger("zip loaded files into" + dir + 'result.zip')

        try:
            zf.write(load_content.result)
            zf.write(load_logo.result)
        finally:
            zf.close()
        zipped = [load_content.result, load_logo.result]
        return zipped


    @task(log_result, zip_result)
    def clear_files(logger):
        logger('clear zipped files')

        for file in zip_result.result:
            os.remove(file)
        pass


    run(clear_files, 2)

output:
::

    build started (log_result, zip_result, load_logo, load_content, create_temp_dir, clear_files)
    task log_result(load_content, load_logo) is added to queue
    task load_content(create_temp_dir) is added to queue
    task create_temp_dir is added to queue
    task create_temp_dir is started
       create_temp_dir:  clear temp dir
    task create_temp_dir is completed in 0 seconds
    task load_logo(create_temp_dir) is added to queue
    task load_logo(create_temp_dir) is started
    task load_logo(create_temp_dir) is completed in 1 seconds
    task load_content(create_temp_dir) is started
    task load_content(create_temp_dir) is completed in 1 seconds
    task zip_result(load_content, load_logo) is added to queue
    task log_result(load_content, load_logo) is started
       log_result:  ./temp/example.py loaded
       log_result:  ./temp/punksta_logo.png loaded
    task log_result(load_content, load_logo) is completed in 0 seconds
    task zip_result(load_content, load_logo) is started
       zip_result:  zip loaded files into./temp/result.zip
    task zip_result(load_content, load_logo) is completed in 0 seconds
    task clear_files(log_result, zip_result) is started
       clear_files:  clear zipped files
    task clear_files(log_result, zip_result) is completed in 0 seconds
    build ended (log_result, zip_result, load_logo, load_content, create_temp_dir, clear_files) in 2 seconds

    Process finished with exit code 0


License
=======

microbuild is licensed under a MIT license. See `LICENSE.txt`_.

.. _LICENSE.txt: https://github.com/CalumJEadie/microbuild/blob/master/LICENSE.txt


CHANGES.rst
Changes
=======

0.4.0 - 08/05/2016
------------------
* Forked from microbuild



0.3.3 - 26/03/2013
------------------

* Improved README documentation.

0.3.1,0.3.2 - 15/12/2012
------------------------

* Improved README documentation.

0.3.0 - 18/09/2012
------------------

* Fixed bug in logging. No longer modifies root logger.
* Added ignore functionality.
* Extended API documentation.

0.2.0 - 29/08/2012
------------------

* Added progress tracking output.
* Added handling of exceptions within tasks.

0.1.0 - 28/08/2012
------------------

* Initial release.
* Added management of dependancies between tasks.
* Added automatic generation of command line interface.


async_tasks/tests/async_tasks.py
import functools
import unittest
from time import sleep

from async_tasks.async_tasks import Logger, run, task, Task
from .. import async_tasks


class TestLogger(Logger):
    def log_in_task(selt, task, obj):
        pass

    def str_task(self, task):
        return ""

    def on_build_end(self, tasks, seconds):
        pass

    def on_end_executing(self, task, seconds):
        pass

    def on_added_to_queue(self, task):
        pass

    def on_build_start(self, tasks):
        pass

    @classmethod
    def str_task_list(cls, tasks):
        return ""

    def on_start_executing(self, task):
        pass


class TestBuildSimple(unittest.TestCase):

    def make_task_delay(number):
        def returnWithWhait(logger):
            sleep(1)
            return number
        return Task(returnWithWhait, [])

    def make_task(number):
        return Task(lambda logger: number, [])

    def test_single_thread_result(self):
        r = range(0, 100)
        numbers = list(map(self.make_task, r))

        @task(*numbers)
        def return_sum(logger):
            return functools.reduce(lambda a, b: a + b.result, numbers, 0)

        @task(return_sum)
        def exit(logger):
            return return_sum.result

        run(exit, 1, TestLogger())
        self.assertEqual(return_sum.result, sum(r))

    def test_multi_thread_result(self):
        r = range(0, 100)
        numbers = list(map(self.make_task_delay(), r))

        @task(*numbers)
        def return_sum(logger):
            return functools.reduce(lambda a, b: a + b.result, numbers, 0)

        @task(return_sum)
        def exit(logger):
            return return_sum.result

        run(exit, 100, TestLogger())
        self.assertEqual(return_sum.result, sum(r))


if __name__ == "__main__":
    unittest.main(verbosity=2)



async_tasks/async_tasks.py
__author__ = "Stanislav Shakirov"
__license__ = "MIT License"
__contact__ = "https://github.com/CalumJEadie/microbuild"

import functools

import concurrent
from concurrent.futures import ThreadPoolExecutor
from datetime import datetime
from threading import Lock
tasks = []


class Logger(object):
    def on_end_executing(self, task, seconds):
        print("task %s is completed in %s seconds" % (self.str_task(task), seconds))

    def on_start_executing(self, task):
        print("task %s is started" % self.str_task(task))

    def on_added_to_queue(self, task):
        print("task %s is added to queue" % self.str_task(task))

    def on_build_start(self, tasks):
        print('build started %s' % self.str_task_list(tasks))

    def on_build_end(self, tasks, seconds):
        print('build ended %s in %s seconds ' % (self.str_task_list(tasks), seconds))

    def log_in_task(selt, task, obj):
        print('   %s:  %s' % (task.name, str(obj)))

    def str_task(self, task):
        return task.name +  self.str_task_list(task.dependencies)

    @classmethod
    def str_task_list(cls, tasks):
        def str_item(index, dep, size):
            if index is 0:
                if index is not size - 1:
                    return '(' + dep.name + ', '
                else:
                    return '(' + dep.name + ')'
            elif index is not size - 1:
                return dep.name + ', '
            else:
                return dep.name + ')'

        return functools.reduce(lambda ite_list, index_dep: ite_list + str_item(index_dep[0], index_dep[1], len(tasks)),
                                enumerate(tasks), '')


class LoggerWrapper(object):

    def __init__(self, logger):
        self.logger = logger
        self.messages = {}
        self.print_lock = Lock()

    def on_start_executing(self, task):
        self.messages[task] = []

    def on_end_executing(self, task, seconds):
        self.log_messages_of_task(task, self.messages[task], seconds)

    def on_build_end(self, tasks, seconds):
        self.logger.on_build_end(tasks, seconds)

    def on_build_start(self, tasks):
        self.logger.on_build_start(tasks)

    def on_added_to_queue(self, task):
        self.logger.on_added_to_queue(task)

    def log_in_task(self, task, obj):
        self.messages[task].append(obj)

    def log_messages_of_task(self, task, objects, seconds):
        with self.print_lock:
            self.logger.on_start_executing(task)
            for o in objects: self.logger.log_in_task(task, o)
            self.logger.on_end_executing(task, seconds)


class Task(object):
    def __init__(self, func, dependencies):
        self.func = func
        self.name = func.__name__
        self.dependencies = dependencies
        self.result = None
        tasks.append(self)

    def __call__(self, *args, **kwargs):
        if self.logger:
            self.logger.on_start_executing(self)
        t = datetime.now()

        logger = lambda s: self.logger.log_in_task(self, s) if self.logger else lambda s: {}
        par = self.func.__code__.co_varnames
        if len(par) > 0 and par[0] is 'logger':
            self.result = self.func.__call__(logger)
        else:
            self.result = self.func.__call__()

        if self.logger:
            self.logger.on_end_executing(self,  (datetime.now() - t).seconds)

        return self.result


    def is_complete(self):
        return self.result is not None


    @classmethod
    def is_task(cls, obj):
        return isinstance(obj, cls)

    def set_future(self, future):
        self.future = future

    def set_logger(self, logger):
        self.logger = logger

    def __str__(self):
        return self.name


def task(*dependencies):
    for d in filter(lambda x: not Task.is_task(x), dependencies):
        raise Exception("%s is not a task. Each dependency should be a task." % d.__str__())
    return lambda fn: Task(fn, dependencies)


def inner_run(current_task, future_tasks, executor, notifer):
    for d in current_task.dependencies:
        if d not in future_tasks:
            notifer.on_added_to_queue(d)
            future_d = inner_run(d, future_tasks, executor, notifer)
            future_tasks.add(future_d)

    for f in concurrent.futures.as_completed(map(lambda t: t.future, current_task.dependencies)):
        f.result()

    if current_task not in future_tasks:
        current_task.set_logger(notifer)
        current_task.future = executor.submit(current_task)
        future_tasks.add(current_task)
    return current_task


def run(task_, workers = 1, logger=Logger()):
    with ThreadPoolExecutor(max_workers=workers) as executor:
        run_on_executor(task_, executor, logger)


def get_all_dependecies(task):
    tasks = []
    for d in task.dependencies:
        tasks.append(d)
        for d2 in get_all_dependecies(d):
            tasks.append(d2)
    s = {}
    for t in tasks:
        s[t] = []
    return s.keys()


def run_on_executor(task_, executor, logger=Logger()):
    logger_wrapper = LoggerWrapper(logger)

    deps = list(get_all_dependecies(task_))
    deps.append(task_)

    logger_wrapper.on_build_start(deps)

    def stater():
        futured_task = inner_run(task_, set(), executor, logger_wrapper)
        return futured_task.future.result()

    t = datetime.now()

    stater()

    logger_wrapper.on_build_end(deps, (datetime.now() - t).seconds)


class TeamCityLogger(Logger):
    @classmethod
    def progress_message(cls, message):
        print("##teamcity[progressMessage '" + message + "']")

    @classmethod
    def progress_start(cls, message):
        print("##teamcity[progressStart '" + message + "']")

    @classmethod
    def progress_end(cls, message):
        print("##teamcity[progressFinish '" + message + "']")

    @classmethod
    def status_message(cls, message, status='NORMAL', error_details=''):
        print("##teamcity[message text='%s' errorDetails='%s' status='%s']" % (message, status, error_details))

    def onBuildStart(self, tasks):
        pass

    def onAddedToQueue(self, task):
        pass

    def onBuildEnd(self, tasks):
        pass

    def onStartExecuting(self, task):
        self.progress_start(str(task))

    def log_in_task(self, task, obj):
        self.progress_message(obj)

    def on_build_start(self, tasks):
        pass

    def on_added_to_queue(self, task):
        pass

    def on_end_executing(self, task, seconds):
        self.progress_end(self.str_task(task))

    def on_start_executing(self, task):
        self.progress_start(self.str_task(task))

    def on_build_end(self, tasks, seconds):
        pass




example.py
import os
import shutil
import zipfile

import requests

from async_tasks.async_tasks import task, run, TeamCityLogger

dir = './temp/'


def load_file(url, name):
    response = requests.get(url, stream=True)
    with open(dir + name, "wb") as handle:
        handle.write(response.content)
    return dir + name


@task()
def create_temp_dir(logger):
    if not os.path.exists(dir):
        logger("create temp dir")
        os.makedirs(dir)
    else:
        logger("clear temp dir")
        shutil.rmtree(dir)
        os.makedirs(dir)


@task(create_temp_dir)
def load_logo():
    file_address = load_file("https://avatars1.githubusercontent.com/u/13404754?v=3&s=460", "punksta_logo.png")
    return file_address


@task(create_temp_dir)
def load_content():
    file_address = load_file("https://raw.githubusercontent.com/punksta/async_tasks/master/example.py", "example.py")
    return file_address


@task(load_content, load_logo)
def log_result(logger):
    for name in [load_content.result, load_logo.result]:
        logger(name + " loaded")


@task(load_content, load_logo)
def zip_result(logger):
    zf = zipfile.ZipFile(dir + 'result.zip', mode='w')

    logger("zip loaded files into" + dir + 'result.zip')

    try:
        zf.write(load_content.result)
        zf.write(load_logo.result)
    finally:
        zf.close()
    zipped = [load_content.result, load_logo.result]
    return zipped


@task(log_result, zip_result)
def clear_files(logger):
    logger('clear zipped files')

    for file in zip_result.result:
        os.remove(file)
    pass


run(clear_files, 2, TeamCityLogger())